File: oracle/app/engine/refdata.py
"""Загрузка справочников техприложения из YAML-конфига.

Все справочники — во внешнем файле reference.yaml, НЕ в коде (§2.14 ТЗ):
при изменении техприложения правится только конфиг. Значения соответствуют
официальному техприложению ЛЦТ-2026 (таблицы 4.1, 4.2, 5.1; разделы 8–9).
"""

from __future__ import annotations

from pathlib import Path
from typing import Optional

import yaml

DEFAULT_REF_PATH = Path(__file__).parent / "reference.yaml"


class RefData:
    """Типизированный доступ к справочникам техприложения."""

    def __init__(self, path: Optional[Path] = None):
        self.path = Path(path) if path else DEFAULT_REF_PATH
        with self.path.open("r", encoding="utf-8") as fh:
            raw = yaml.safe_load(fh)
        # Таблица 4.1, отсортированная по возрастанию Ду
        self.diameters = sorted(raw["diameters"], key=lambda d: d["dn_mm"])
        # Таблица 4.2: габариты по Ду
        self.envelopes = {int(e["dn_mm"]): e for e in raw["envelopes"]}
        # Таблица 5.1
        self.restrictions = raw.get("restrictions", {})
        self.tariffs = raw["tariffs_rub"]
        self.scoring = raw["scoring"]
        self.rules = raw["rules"]
        self.depth = raw.get("depth", {})

    # ---- диаметры / гидравлика (таблица 4.1) ----

    def diameter_for_flow(self, flow_tph: float) -> dict:
        """Минимальный Ду, пропускная способность которого покрывает расход."""
        for d in self.diameters:
            if d["max_flow_tph"] >= flow_tph:
                return d
        return self.diameters[-1]  # берём максимальный; факт нехватки — в warnings

    def diameter_entry(self, dn_mm: float) -> dict:
        for d in self.diameters:
            if d["dn_mm"] == dn_mm:
                return d
        return self.diameters[-1]

    def next_diameter(self, dn_mm: float) -> dict:
        """Следующий больший Ду (для повышения при превышении предельной длины)."""
        bigger = [d for d in self.diameters if d["dn_mm"] > dn_mm]
        return bigger[0] if bigger else self.diameters[-1]

    def max_len_for(self, dn_mm: float) -> float:
        return float(self.diameter_entry(dn_mm)["max_len_m"])

    # ---- габариты (таблица 4.2) ----

    def envelope_width_m(self, dn_mm: float) -> float:
        return float(self.envelopes.get(int(dn_mm), self.envelopes[max(self.envelopes)])["width_m"])

    def envelope_height_m(self, dn_mm: float) -> float:
        return float(self.envelopes.get(int(dn_mm), self.envelopes[max(self.envelopes)])["height_m"])

    # ---- ограничения (таблица 5.1) ----

    def restriction_rule(self, restriction_type: str) -> Optional[dict]:
        return self.restrictions.get(restriction_type)

    def min_distance_m(self, restriction_type: str, dn_mm: float = 0) -> float:
        """Минимальное горизонтальное расстояние; для ОКС — ступенями по Ду."""
        rule = self.restrictions.get(restriction_type) or {}
        if "min_distance_by_dn" in rule:
            for step in rule["min_distance_by_dn"]:
                if dn_mm <= step["max_dn_mm"]:
                    return float(step["distance_m"])
            return float(rule["min_distance_by_dn"][-1]["distance_m"])
        return float(rule.get("min_distance_m", 1.0))

    # ---- тарифы (раздел 8) ----

    def lay_tariff(self, dn_mm: float) -> float:
        """Ставка нового строительства, ₽/м (таблица 4.1)."""
        return float(self.diameter_entry(dn_mm)["cost_new_rub_m"])

    def recon_tariff(self, dn_mm: float) -> float:
        """Ставка реконструкции по ТРЕБУЕМОМУ диаметру, ₽/м (§7, таблица 4.1)."""
        return float(self.diameter_entry(dn_mm)["cost_recon_rub_m"])

    def tie_in_cost(self) -> float:
        """§8.2: стоимость одной врезки (любого вида)."""
        return float(self.tariffs["tie_in"])

    def chamber_cost(self, max_dn_mm: float) -> float:
        """§8.2: стоимость камеры по наибольшему Ду примыкающих участков."""
        for step in self.tariffs["chamber_cost_scale"]:
            if max_dn_mm <= step["max_dn_mm"]:
                return float(step["cost_rub"])
        return float(self.tariffs["chamber_cost_scale"][-1]["cost_rub"])

    def unconnected_penalty(self, flow_tph: float) -> float:
        """§8.3: штраф за неподключённый ОКС = 100 млн + 500 тыс. × G."""
        return float(self.tariffs["unconnected_penalty_base"]) \
            + float(self.tariffs["unconnected_penalty_per_tph"]) * flow_tph

    def tariff(self, key: str) -> float:
        return float(self.tariffs[key])

    def rule(self, key: str) -> float:
        return float(self.rules[key])

    # ---- ранжирование (раздел 9) ----

    def score(self, calculated_cost_rub: float, length_m: float) -> float:
        """S = w_c·C/C0 + w_l·L/L0; меньше — лучше."""
        s = self.scoring
        return (float(s["weight_cost"]) * calculated_cost_rub / float(s["cost_norm_rub"])
                + float(s["weight_length"]) * length_m / float(s["length_norm_m"]))

    # ---- задание на глубину (раздел 6, доп. задача) ----

    def depth_rule(self, key: str, default: float) -> float:
        return float(self.depth.get(key, default))

    def depth_k(self, depth_m: float) -> float:
        """Kгл = 1 + rate·(h − free) при h > free, иначе 1 (без скидки за мельче)."""
        free = self.depth_rule("free_depth_m", 3.0)
        if depth_m <= free:
            return 1.0
        return 1.0 + self.depth_rule("depth_cost_rate", 0.10) * (depth_m - free)

File: app/src/main/java/ru/heatnet/dit/engine/core/RefData.java
package ru.heatnet.dit.engine.core;

import org.yaml.snakeyaml.Yaml;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Загрузка справочников техприложения из YAML-конфига.
 *
 * Все справочники — во внешнем файле reference.yaml, НЕ в коде (§2.14 ТЗ).
 * Источник файла: ENGINE_REFERENCE_PATH, иначе ресурс classpath
 * /reference.yaml (упакован в jar).
 */
public class RefData {

    public static final String DEFAULT_RESOURCE = "/reference.yaml";

    // Таблица 4.1, по возрастанию Ду
    public final List<Map<String, Object>> diameters = new ArrayList<>();
    // Таблица 4.2: габариты по Ду
    public final Map<Integer, Map<String, Object>> envelopes = new LinkedHashMap<>();
    // Таблица 5.1
    public final Map<String, Object> restrictions;
    public final Map<String, Object> tariffs;
    public final Map<String, Object> scoring;
    public final Map<String, Object> rules;
    public final Map<String, Object> depth;

    @SuppressWarnings("unchecked")
    public RefData() {
        Map<String, Object> raw = loadYaml();
        List<Map<String, Object>> ds = (List<Map<String, Object>>) raw.get("diameters");
        ds.sort(Comparator.comparingDouble(d -> num(d.get("dn_mm"))));
        diameters.addAll(ds);
        for (Map<String, Object> e : (List<Map<String, Object>>) raw.get("envelopes")) {
            envelopes.put((int) num(e.get("dn_mm")), e);
        }
        restrictions = (Map<String, Object>) raw.getOrDefault("restrictions", new LinkedHashMap<>());
        tariffs = (Map<String, Object>) raw.get("tariffs_rub");
        scoring = (Map<String, Object>) raw.get("scoring");
        rules = (Map<String, Object>) raw.get("rules");
        depth = (Map<String, Object>) raw.getOrDefault("depth", new LinkedHashMap<>());
    }

    private static Map<String, Object> loadYaml() {
        String override = System.getenv("ENGINE_REFERENCE_PATH");
        try (InputStream in = override != null && !override.isEmpty()
                ? Files.newInputStream(Paths.get(override))
                : RefData.class.getResourceAsStream(DEFAULT_RESOURCE)) {
            if (in == null) {
                throw new IllegalStateException("не найден " + DEFAULT_RESOURCE + " в classpath");
            }
            return new Yaml().load(in);
        } catch (IOException e) {
            throw new IllegalStateException("не удалось прочитать reference.yaml: " + e.getMessage(), e);
        }
    }

    static double num(Object v) {
        return ((Number) v).doubleValue();
    }

    // ---- диаметры / гидравлика (таблица 4.1) ----

    /** Минимальный Ду, пропускная способность которого покрывает расход. */
    public Map<String, Object> diameterForFlow(double flowTph) {
        for (Map<String, Object> d : diameters) {
            if (num(d.get("max_flow_tph")) >= flowTph) {
                return d;
            }
        }
        return diameters.get(diameters.size() - 1); // факт нехватки — в warnings
    }

    public Map<String, Object> diameterEntry(double dnMm) {
        for (Map<String, Object> d : diameters) {
            if (num(d.get("dn_mm")) == dnMm) {
                return d;
            }
        }
        return diameters.get(diameters.size() - 1);
    }

    /** Следующий больший Ду (для повышения при превышении предельной длины). */
    public Map<String, Object> nextDiameter(double dnMm) {
        for (Map<String, Object> d : diameters) {
            if (num(d.get("dn_mm")) > dnMm) {
                return d;
            }
        }
        return diameters.get(diameters.size() - 1);
    }

    public double maxLenFor(double dnMm) {
        return num(diameterEntry(dnMm).get("max_len_m"));
    }

    // ---- габариты (таблица 4.2) ----

    private Map<String, Object> envelope(double dnMm) {
        Map<String, Object> e = envelopes.get((int) dnMm);
        if (e == null) {
            e = envelopes.get(envelopes.keySet().stream().mapToInt(Integer::intValue).max().orElse(0));
        }
        return e;
    }

    public double envelopeWidthM(double dnMm) {
        return num(envelope(dnMm).get("width_m"));
    }

    public double envelopeHeightM(double dnMm) {
        return num(envelope(dnMm).get("height_m"));
    }

    // ---- ограничения (таблица 5.1) ----

    @SuppressWarnings("unchecked")
    public Map<String, Object> restrictionRule(String restrictionType) {
        return (Map<String, Object>) restrictions.get(restrictionType);
    }

    /** Минимальное горизонтальное расстояние; для ОКС — ступенями по Ду. */
    @SuppressWarnings("unchecked")
    public double minDistanceM(String restrictionType, double dnMm) {
        Map<String, Object> rule = restrictionRule(restrictionType);
        if (rule == null) {
            return 1.0;
        }
        Object byDn = rule.get("min_distance_by_dn");
        if (byDn != null) {
            List<Map<String, Object>> steps = (List<Map<String, Object>>) byDn;
            for (Map<String, Object> step : steps) {
                if (dnMm <= num(step.get("max_dn_mm"))) {
                    return num(step.get("distance_m"));
                }
            }
            return num(steps.get(steps.size() - 1).get("distance_m"));
        }
        return num(rule.getOrDefault("min_distance_m", 1.0));
    }

    // ---- тарифы (раздел 8) ----

    /** Ставка нового строительства, ₽/м (таблица 4.1). */
    public double layTariff(double dnMm) {
        return num(diameterEntry(dnMm).get("cost_new_rub_m"));
    }

    /** §3.2: стоимость врезки в существующую камеру (за новый участок). */
    public double tieInCost() {
        return num(tariffs.get("tie_in_existing_chamber"));
    }

    /** §3.2: стоимость камеры по наибольшему ДУ примыкающих участков. */
    @SuppressWarnings("unchecked")
    public double chamberCost(double maxDnMm) {
        List<Map<String, Object>> scale = (List<Map<String, Object>>) tariffs.get("chamber_cost_scale");
        for (Map<String, Object> step : scale) {
            if (maxDnMm <= num(step.get("max_dn_mm"))) {
                return num(step.get("cost_rub"));
            }
        }
        return num(scale.get(scale.size() - 1).get("cost_rub"));
    }

    /** §6: штраф за неподключённую точку = 100 млн + 500 тыс. × G. */
    public double unconnectedPenalty(double flowTph) {
        return num(tariffs.get("unconnected_penalty_base"))
                + num(tariffs.get("unconnected_penalty_per_tph")) * flowTph;
    }

    public double rule(String key) {
        return num(rules.get(key));
    }

    // ---- ранжирование (раздел 9) ----

    /** S = w_c·C/C0 + w_l·L/L0; меньше — лучше. */
    public double score(double calculatedCostRub, double lengthM) {
        return num(scoring.get("weight_cost")) * calculatedCostRub / num(scoring.get("cost_norm_rub"))
                + num(scoring.get("weight_length")) * lengthM / num(scoring.get("length_norm_m"));
    }

    // ---- задание на глубину (раздел 6, доп. задача) ----

    public double depthRule(String key, double defaultValue) {
        Object v = depth.get(key);
        return v == null ? defaultValue : num(v);
    }

    /** Kгл = 1 + rate·(h − free) при h > free, иначе 1 (без скидки за мельче). */
    public double depthK(double depthM) {
        double free = depthRule("free_depth_m", 3.0);
        if (depthM <= free) {
            return 1.0;
        }
        return 1.0 + depthRule("depth_cost_rate", 0.10) * (depthM - free);
    }

    /**
     * §4: условный габарит и глубина существующей коммуникации
     * (глубина до ВЕРХА габарита; габарит heat_network — по таблице 1).
     * null — тип не описан.
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> existingUtility(String restrictionType) {
        Object m = depth.get("existing_utilities");
        if (!(m instanceof Map)) {
            return null;
        }
        return (Map<String, Object>) ((Map<String, Object>) m).get(restrictionType);
    }
}
